frontend/lib/features/pfis/pfi_providers.dart

```diff
@@ -1,22 +1,56 @@
+import 'dart:convert';
 import 'package:flutter_starter/features/pfis/pfi.dart';
 import 'package:hooks_riverpod/hooks_riverpod.dart';
+import 'package:http/http.dart' as http;
+import 'package:shared_preferences/shared_preferences.dart';
 
-final pfisProvider = Provider<List<Pfi>>(
-  (ref) => [
-    Pfi(
-      id: 'prototype',
-      name: 'Prototype',
-      didUri: 'did:dht:74hg1efatndi8enx3e4z6c4u8ieh1xfkyay4ntg4dg1w6risu35y',
-    ),
-    Pfi(
-      id: 'africa',
-      name: 'Africa',
-      didUri: 'coming soon...',
-    ),
-    Pfi(
-      id: 'mexico',
-      name: 'Mexico',
-      didUri: 'coming soon...',
-    ),
-  ],
-);
+final pfisProvider = FutureProvider<List<Pfi>>((ref) async {
+  const url = 'https://raw.githubusercontent.com/TBD54566975/pfi-providers-data/main/pfis.json';
+  const cacheKey = 'pfis_cache';
+
+  try {
+    // Attempt to fetch data from the URL
+    final response = await http.get(Uri.parse(url));
+    if (response.statusCode == 200) {
+      // Parse the JSON data directly into a list of Pfi objects
+      List<Pfi> pfis = (json.decode(response.body) as List).map((data) {
+        return Pfi(
+          id: data['id'] as String,
+          name: data['name'] as String,
+          didUri: data['didUri'] as String,
+        );
+      }).toList();
+
+      // Save the data to cache
+      final prefs = await SharedPreferences.getInstance();
+      await prefs.setString(cacheKey, response.body);
+
+      return pfis;
+    } else {
+      // If server returns an unsuccessful response, try loading from cache
+      return await _loadFromCache(cacheKey);
+    }
+  } catch (e) {
+    // In case of an error, try loading from cache
+    return await _loadFromCache(cacheKey);
+  }
```

Comment:
Another approach I like to use on mobile, to keep things snappy is to always load cache first, then always load from network. That way the users gets the list right away from cache and we can refresh from network after. This would require another provider/notifier which is a bit more complicated. Happy to show how that might work as well if you're interested.

Reply:
good idea - did that.